### ISSUE TITLE
training speed and abmil.

### ISSUE THREAD

**Issue by Jiawei-Yang** [binli123/dsmil-wsi#27]
Hi, thanks for your fabulous project.

May we get some information from your experience on the training speed of each part of the pipeline? e.g., how long does pre-training on C16/TCGA take? and how long does the MIL part take?

For example, I'm currently running the `train_tcga.py` using the provided C16 features. But the training speed seems to be low and the GPU usage utility also stays low. Is there any way to speed up?

-----
About ABMIL

I'm wondering if you planned to release the ABMIL code?

**Comment by binli123** [comment]
> Hi, thanks for your fabulous project.
> 
> May we get some information from your experience on the training speed of each part of the pipeline? e.g., how long does pre-training on C16/TCGA take? and how long does the MIL part take?
> 
> For example, I'm currently running the `train_tcga.py` using the provided C16 features. But the training speed seems to be low and the GPU usage utility also stays low. Is there any way to speed up?
> 
> About ABMIL
> 
> I'm wondering if you planned to release the ABMIL code?

It can take up to several hours, depending on which GPU you use for training. The low usage is because the batch size is 1. Parallelizing MIL models with batch sizes bigger than 1 requires quite a bit of coding since the bags have different lengths and putting them into a batch is not natively supported by PyTorch. 

The contrastive learning pipeline takes a significantly longer time to train, in my case, about 2 weeks on a GPU cluster of 4 1080TIs

For ABMIL you could easily adapt code from the author's implementation on GitHub: https://github.com/AMLab-Amsterdam/AttentionDeepMIL

**Comment by Jiawei-Yang** [comment]
Thanks for your quick response!

I'm currently training the MIL part using the provided C16 features. It seems to take about 20 hours or more for 200-epoch training. Is that normal? or my speed is much slower than the normal case?

In my machine, the bottleneck of the code seems to be at the data IO and the hard parallelizing dilemma.

Again, thanks for your wonderful project and kind response! Your reply helps a lot.

**Comment by binli123** [comment]
The training speed has now improved significantly with the latest commit. I also incorporated the training/testing into the same pipeline in the latest commit. This change allows you to read the evaluation results on a reserved test set. I also incorporated a simple weights initialization method which helps stabilize the training. You can set --eval_scheme=5-fold-cv-standalone-test which will perform a train/valid/test like this: 
>A standalone test set consisting of 20% samples is reserved, remaining 80% of samples are used to construct a 5-fold cross-validation.
>For each fold, the best model and corresponding threshold are saved.
>After the 5-fold cross-validation, 5 best models along with the corresponding optimal thresholds are obtained which are used to perform inference on the reserved test set. A final prediction for a test sample is the majority vote of the 5 models.
>For a binary classification, accuracy and balanced accuracy scores are computed. For a multi-label classification, hamming loss (smaller the better) and subset accuracy are computed.

You can also simply run a 5-fold cv --eval_scheme=5-fold-cv

There were some issues with the testing script when loading pretrained weights (i.e., sometimes the weights are not fully loaded or there are missing weights, setting strict=False can reveal the problems.). The purpose of the testing script is to generate the heatmap, you should now read the performance directly from the training script. I will fix the issues in a couple of days.
